README.md
# Bidoo's plugins for [VCVRack](https://vcvrack.com)

<!-- Version and License Badges -->
![Version](https://img.shields.io/badge/version-2.1.1-green.svg?style=flat-square)
![License](https://img.shields.io/badge/license-BSD3-blue.svg?style=flat-square)
![Language](https://img.shields.io/badge/language-C++-yellow.svg?style=flat-square)

<p align="center">
   <img src="https://github.com/sebastien-bouffier/Bidoo/blob/v2/images/bidooV2.png"/>
</p>

## How to

You can find information on that plugins pack in the [wiki](https://github.com/sebastien-bouffier/Bidoo/wiki). When doing tests it happens that I record a video so you may find some ideas on how to use those modules [here](https://www.youtube.com/bidoo).

## License

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

All **graphics** in the `res` directory are copyright © 2019 Sébastien Bouffier (sebastien.bouffier@gmail.com) and licensed under [CC BY-NC-ND 4.0](https://creativecommons.org/licenses/by-nc-nd/4.0/).

## Donate

If you enjoy those modules you can support the development by making a donation. Here's the link: [DONATE](https://paypal.me/sebastienbouffier)


changelog.md
### v2.1.1 (06/02/2023)
- zOù MAï and enCORE can copy/paste/randomize/erase per page thanks to Tavis

### v2.1.0 (12/01/2023)
- ANTn minimp3 config change for SIMD optimizations

### v2.0.29 (24/11/2022)
- ANTn minimp3 config change for OSX build compability

### v2.0.28 (22/11/2022)
- cANARd fix zoom issue

### v2.0.27 (30/10/2022)
- πlOT copy/paste scenes available on top and bottom scene buttons (can be applied between banks)
- πlOT controls behavior can be switched between "constant override" (default mode) i.e. it keeps its value until a scene is saved or the morph button is clicked, and not "constant override" meaning that the control jumps to its value when changing scene.

### v2.0.26 (11/10/2022)
- fix Zou triggers issue

### v2.0.25 (9/10/2022)
- ratEaU is available

### v2.0.24 (1/10/2022)
- πlOT fix OSX build

### v2.0.23 (1/10/2022)
- πlOT fix on gate mode
- rabBIT randomize implementation

### v2.0.22 (15/09/2022)
- πlOT can add masking tape on top of controls in order to display labels (use menu)

### v2.0.21 (07/09/2022)
- mem leaks fixes thanks to Filipe

### v2.0.20 (24/08/2022)
- pitchshifters sample rate change fix thanks to Filipe
- C++ fixes thanks to Filipe

### v2.0.19 (10/08/2022)
- all samplers are not case sensitive anymore on files extensions
- πlOT bank copy/paste capability
- enCORE fix PRE/!PRE probability issue
- antN better .pls .m3u links management
- dIKT@T root note on channels > 1 issue fixed
- cANARd sample position can be automated via index input in loop mode
- EMILE crash VCV under OS X when deleting module with an image loaded ... fixed

### v2.0.17 (21/07/2022)
- tOCAnTe upgrade for more divisions + dedicated output for enCORE sequencer
- DIKTAT some scales/modes added (all modules with quantization benefit from that upgrade)
- enCORE sequencer is a sibling of zOù MAï the difference is the clock management since enCORE needs 32 clock pulses per step so it is less precise in terms of step length/trim/retrig etc... but it does not suffer of bpm calculation changes on startup between the three first clock ticks.

### v2.0.16 (09/07/2022)
- zOù MAï pattern CV input rescaled

### v2.0.15 (06/07/2022)
- quantization issue in all modules fix
- zOù MAï track naming capability via menu

### v2.0.14 (19/05/2022)
- bordL random skip and slide fix
- liMonADe wave table and frame wav export, first step for wave table png export
- zOù MAï run button added
- zOù MAï CV1 and 2 range is now -10V +10V

### v2.0.13 (08/04/2022)
- changelog link update

### v2.0.12 (06/04/2022)
- rename VOID module class

### v2.0.10 (03/04/2021)
- liMonADe scan bug fixed when table is empty

### v2.0.9 (21/03/2022)
- last theme applied becomes the new default theme (modules are drawn accordingly in the browser once the application is restarted)
- bordL's skip, slide, play mode and count mode buttons should be OSC compliant
- bordL has some perf improvements

### v2.0.8 (27/01/2022)
- bAr and mINIBar attack and release params where inverted
- all modules have themes (light, dark, black, blue and green)

### v2.0.7 (08/01/2022)
- zOù MAï sync issues fixed
- zOù MAï VO quantization per track
- zOù MAï CV1 can be quantized too
- zOù MAï slide timing vs. rate constant per track
- zOù MAï slide base calculation per note, note length vs. 1 step length
- zOù MAï tools removed from interface, tools extended and moved in menu and accessible by "over" shortcuts (description/mapping in menu)
- zOù MAï expander with new features (each pattern has its own set of params) :
- zOù MAï Fill per track
- zOù MAï Force all rand steps per track
- zOù MAï Mute all rand steps (except dice  100% i.e. no rand) per track
- zOù MAï Transpose per track (0+10V or -4+6V base)
- zOù MAï Increase or decrease dice value for steps with rand dice less than 100% per track
- zOù MAï Rotate pattern left or right per track (rotate shift and length as params)

### v1.1.39 (26/09/2021)
- πlOT Orange mode for controls => Like in Red mode they are quantized in order to drive a VCO but they jump from note to note like in Yellow mode.
- πlOT sequencer can record controls values

### v1.1.38 (26/09/2021)
- πlOT gate time issue fixed
- πlOT sequencer mode 3 => negative snapshots fixed

### v1.1.37 (25/09/2021)
- πlOT UI redesigned
- πlOT red mode for controls = quantization to semi-tones + display of note when edited
- πlOT saving a scene release all the controls
- πlOT red wait mode for the sequencer = morph must reach min or max to change controls in yellow mode (jump) and to fire the gates of controls in green mode (gate)
- πlOT yellow wait mode for the sequencer = morph changes controls in yellow mode (jump) when a trig occurs and fires the gates of controls in green mode (gate)
- πlOT bezier curve morphing error fixed

### v1.1.36 (21/09/2021)
- πlOT custom morphing curve

### v1.1.32 (21/09/2021)
- MOiRE random button for target scene added
- πlOT is a kind of MOiRE++ with automorphing between scenes and more

### v1.1.31 (11/09/2021)
- maGma empty sample bug fix
- póUPrè empty sample bug fix

### v1.1.30 (30/05/2021)
- ACnE stereo link bug fixed

### v1.1.29 (03/03/2021)
- zOù MAï Sync issue on Reset and Pattern change fixed
- bordL & dTrOY lydian scale quantization fixed

### v1.1.28 (22/10/2020)
- ACnE ramp fix between snaphots
- ACnE ramp is applied on Solo and Mute
- zOù MAï run input reset clock fix

### v1.1.27 (13/10/2020)
- tOCAnTe reset and run outputs added
- zOù MAï run input

### v1.1.26 (02/10/2020)
- bAr and mINIBar compressor tag
- new tape stop module fREIN

### v1.1.25 (14/09/2020)
- zOù MAï latched buttons
- zOù MAï record V/O + gate capability

### v1.1.24 (12/09/2020)
- mINIBar shows side chain and input signals at the same time
- zOù MAï latched octave buttons

### v1.1.23 (12/09/2020)
- per-module manual URLs
- Res input added on FFilTr

src/ANTN.cpp
#include "plugin.hpp"
#include "dsp/digital.hpp"
#include "dsp/resampler.hpp"
#include "BidooComponents.hpp"
#include "dsp/ringbuffer.hpp"
#include <algorithm>
#include <cctype>
#include <atomic>
#include <sstream>
#include <thread>
#include "curl/curl.h"
#define MINIMP3_IMPLEMENTATION
#include "dep/minimp3/minimp3.h"

using namespace std;

const int DATASIZE = 220000000;
const int AUDIOSIZE = 524288;

struct threadReadData {
  dsp::DoubleRingBuffer<char,DATASIZE> *dataToDecodeRingBuffer;
  std::string url;
  std::string secUrl;
  std::atomic<bool> *dl;
  std::atomic<bool> *free;
  float sr;
};

struct threadDecodeData {
  dsp::DoubleRingBuffer<char,DATASIZE> *dataToDecodeRingBuffer;
  dsp::DoubleRingBuffer<dsp::Frame<2>,AUDIOSIZE> *dataAudioRingBuffer;
  mp3dec_t mp3d;
  std::atomic<bool> *dc;
  std::atomic<bool> *free;
  float sr;
};

size_t WriteMemoryCallback(void *contents, size_t size, size_t nmemb, void *userp)
{
  struct threadReadData *pData = (struct threadReadData *) userp;
  size_t realsize = size * nmemb;
  if ((pData->dl->load()) && (realsize < pData->dataToDecodeRingBuffer->capacity()))
  {
    memcpy(pData->dataToDecodeRingBuffer->endData(), contents, realsize);
    pData->dataToDecodeRingBuffer->endIncr(realsize);
    return realsize;
  }
  return 0;
}

size_t WriteUrlCallback(void *contents, size_t size, size_t nmemb, void *userp)
{
  struct threadReadData *pData = (struct threadReadData *) userp;
  size_t realsize = size * nmemb;
  pData->secUrl += (const char*) contents;
  return realsize;
}

void * threadDecodeTask(threadDecodeData data)
{
  data.free->store(false);
  mp3dec_frame_info_t info;
  dsp::DoubleRingBuffer<dsp::Frame<2>,AUDIOSIZE> *tmpBuffer = new dsp::DoubleRingBuffer<dsp::Frame<2>,AUDIOSIZE>();
  dsp::SampleRateConverter<2> conv;
  int inSize;
  int outSize;
  while (data.dc->load()) {
    short pcm[MINIMP3_MAX_SAMPLES_PER_FRAME];
    if ((data.dataToDecodeRingBuffer->size() >= 65536) && (data.dataAudioRingBuffer->capacity() >= 65536)) {
      int samples = mp3dec_decode_frame(&data.mp3d, (const uint8_t*)data.dataToDecodeRingBuffer->startData(), data.dataToDecodeRingBuffer->size(), pcm, &info);

      if (info.frame_bytes > 0) {
        data.dataToDecodeRingBuffer->startIncr(info.frame_bytes);
      }

      if (samples > 0) {
        if (info.channels == 1) {
          for(int i = 0; i < samples; i++) {
            if (!data.dc->load()) break;
            dsp::Frame<2> newFrame;
            newFrame.samples[0]=(float)pcm[i] * 30517578125e-15f;
            newFrame.samples[1]=(float)pcm[i] * 30517578125e-15f;
            tmpBuffer->push(newFrame);
          }
        }
        else {
          for(int i = 0; i < 2 * samples; i=i+2) {
            if (!data.dc->load()) break;
            dsp::Frame<2> newFrame;
            newFrame.samples[0]=(float)pcm[i] * 30517578125e-15f;
            newFrame.samples[1]=(float)pcm[i+1] * 30517578125e-15f;
            tmpBuffer->push(newFrame);
          }
        }

        conv.setRates(info.hz, data.sr);
        conv.setQuality(10);
        inSize = tmpBuffer->size();
        outSize = data.dataAudioRingBuffer->capacity();
        conv.process(tmpBuffer->startData(), &inSize, data.dataAudioRingBuffer->endData(), &outSize);
        tmpBuffer->startIncr(inSize);
        data.dataAudioRingBuffer->endIncr((size_t)outSize);
      }
    }
  }
  data.free->store(true);
  return 0;
}

void * threadReadTask(threadReadData data)
{
  data.free->store(false);


  std::string url;
  data.secUrl == "" ? url = data.url : url = data.secUrl;

  if ((rack::system::getExtension(data.url) == ".pls") || (rack::system::getExtension(data.url) == ".m3u")) {
    istringstream iss(url);
    for (std::string line; std::getline(iss, line); )
    {
      std::size_t found=line.find("http");
      if (found!=std::string::npos) {
        url = line.substr(found);
        url.erase(std::remove_if(url.begin(), url.end(), [](unsigned char x){return std::isspace(x);}), url.end());
        url = url.substr(0, url.find("?", 0));
        break;
      }
    }
  }

  std::cout << "URL" << '\n';
  std::cout << data.url << '\n';
  std::cout << "secURL" << '\n';
  std::cout << data.secUrl << '\n';
  std::cout << "url" << '\n';
  std::cout << url << '\n';


  if ((rack::system::getExtension(url) == ".pls") || (rack::system::getExtension(url) == ".m3u")) {
    CURL *curl;
    curl = curl_easy_init();
    data.url=url;
    data.secUrl="";
    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(curl, CURLOPT_URL, data.url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteUrlCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &data);
    curl_easy_perform(curl);
    curl_easy_cleanup(curl);
    data.free->store(true);
    thread iThread = thread(threadReadTask, data);
    iThread.detach();
  }
  else {
    CURL *curl;
    curl = curl_easy_init();
    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
    curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
    curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
    curl_easy_setopt(curl, CURLOPT_URL, url.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &data);
    curl_easy_perform(curl);
    curl_easy_cleanup(curl);
    data.free->store(true);
  }

  return 0;
}


void * urlTask(threadReadData data)
{
  data.free->store(false);

  CURL *curl;
  curl = curl_easy_init();
  curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
  curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
  curl_easy_setopt(curl, CURLOPT_FOLLOWLOCATION, 1L);
  curl_easy_setopt(curl, CURLOPT_URL, data.url.c_str());
  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteUrlCallback);
  curl_easy_setopt(curl, CURLOPT_WRITEDATA, &data);
  data.secUrl = "";
  curl_easy_perform(curl);
  curl_easy_cleanup(curl);

  data.free->store(true);

  thread iThread = thread(threadReadTask, data);
  iThread.detach();

  return 0;
}

struct ANTN : BidooModule {
	enum ParamIds {
		URL_PARAM,
		TRIG_PARAM,
    GAIN_PARAM,
		NUM_PARAMS
	};
	enum InputIds {
		NUM_INPUTS
	};
	enum OutputIds {
		OUTL_OUTPUT,
		OUTR_OUTPUT,
		NUM_OUTPUTS
	};
	enum LightIds {
    ON_LIGHT,
		NUM_LIGHTS = ON_LIGHT+3
	};

  std::string url;
	dsp::SchmittTrigger trigTrigger;
  bool read = false;
  dsp::DoubleRingBuffer<dsp::Frame<2>,AUDIOSIZE> dataAudioRingBuffer;
  dsp::DoubleRingBuffer<char,DATASIZE> dataToDecodeRingBuffer;
  thread rThread, dThread;
  threadReadData rData;
  threadDecodeData dData;
  std::atomic<bool> tDl;
  std::atomic<bool> tDc;
  std::atomic<bool> trFree;
  std::atomic<bool> tdFree;
  bool dirty = false;

	ANTN() {
    config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);
		configParam(GAIN_PARAM, 0.0f, 3.f, 1.f, "Gain");
    configParam(TRIG_PARAM, 0.f, 1.f, 0.f, "Trig");

    tDl.store(true);
    tDc.store(true);
    trFree.store(true);
    tdFree.store(true);

    rData.dataToDecodeRingBuffer = &dataToDecodeRingBuffer;
    rData.dl = &tDl;
    rData.free = &trFree;

    dData.dataToDecodeRingBuffer = &dataToDecodeRingBuffer;
    dData.dataAudioRingBuffer = &dataAudioRingBuffer;
    dData.dc = &tDc;
    dData.free = &tdFree;
    mp3dec_init(&dData.mp3d);
	}

  ~ANTN() {
    tDc.store(false);
    while(!tdFree) {
    }

    tDl.store(false);
    while(!trFree) {
    }
  }

  json_t *dataToJson() override {
    json_t *rootJ = BidooModule::dataToJson();
    json_object_set_new(rootJ, "url", json_string(url.c_str()));
    return rootJ;
  }

  void dataFromJson(json_t *rootJ) override {
    BidooModule::dataFromJson(rootJ);
    json_t *urlJ = json_object_get(rootJ, "url");
  	if (urlJ)
  		url = json_string_value(urlJ);
    dirty = true;
  }

	void process(const ProcessArgs &args) override;

  void onSampleRateChange() override;

  void onReset() override {
		url = "";
		dirty = true;
	}
};

void ANTN::onSampleRateChange() {
  read = false;
  dataAudioRingBuffer.clear();
}

void ANTN::process(const ProcessArgs &args) {
	if (trigTrigger.process(params[TRIG_PARAM].value)) {
    if (read) {
      tDc.store(false);
      while(!tdFree) {
      }

      tDl.store(false);
      while(!trFree) {
      }
      read = false;
      dataToDecodeRingBuffer.clear();
      dataAudioRingBuffer.clear();
    }
    else {
      tDl.store(true);
      rData.url = url;
      rData.sr = args.sampleRate;
      if ((rack::system::getExtension(rData.url) == ".m3u") || (rack::system::getExtension(rData.url) == ".pls")) {
        rThread = thread(urlTask, std::ref(rData));
      }
      else {
        rThread = thread(threadReadTask, std::ref(rData));
      }
      rThread.detach();

      tDc.store(true);
      dData.sr = args.sampleRate;
      mp3dec_init(&dData.mp3d);
      dThread = thread(threadDecodeTask, std::ref(dData));
      dThread.detach();
    }
	}

  if ((dataAudioRingBuffer.size()>64000) && (args.sampleRate<96000)) {
    read = true;
  }
  if ((dataAudioRingBuffer.size()>128000) && (args.sampleRate>=96000)) {
    read = true;
  }

  lights[ON_LIGHT].setBrightness(read ? 0.0f : 1.0f);
  lights[ON_LIGHT+1].setBrightness(read ? (dataAudioRingBuffer.size()<64000 ? 1.0f : 0.0f) : 0.0f);
  lights[ON_LIGHT+2].setBrightness(read ? 1.0f : 0.0f);

  if (read) {
    dsp::Frame<2> currentFrame = *dataAudioRingBuffer.startData();
    outputs[OUTL_OUTPUT].setVoltage(5.0f*currentFrame.samples[0]*params[GAIN_PARAM].getValue());
    outputs[OUTR_OUTPUT].setVoltage(5.0f*currentFrame.samples[1]*params[GAIN_PARAM].getValue());
    dataAudioRingBuffer.startIncr(1);
  }
}

struct ANTNTextField : LedDisplayTextField {
  ANTN *module;

	void step() override {
		LedDisplayTextField::step();
		if (module && module->dirty) {
			setText(module->url);
			module->dirty = false;
		}
	}

	void onChange(const ChangeEvent& e) override {
		if (module && getText().size()>0)
    {
      std::string tText = getText();
      tText.erase(std::remove_if(tText.begin(), tText.end(), [](unsigned char x){return std::isspace(x);}), tText.end());
      module->url = tText;
    }
	}
};

struct ANTNDisplay : LedDisplay {
	void setModule(ANTN* module) {
		ANTNTextField* textField = createWidget<ANTNTextField>(Vec(0, 0));
		textField->box.size = box.size;
		textField->multiline = true;
		textField->module = module;
		addChild(textField);
	}
};

struct ANTNBufferDisplay : TransparentWidget {
	ANTN *module;

	ANTNBufferDisplay() {

	}

  void drawLayer(const DrawArgs& args, int layer) override {
  	if (layer == 1) {
      if (module) {
        nvgSave(args.vg);
      	nvgStrokeWidth(args.vg, 1.0f);
        nvgStrokeColor(args.vg, BLUE_BIDOO);
        nvgFillColor(args.vg, BLUE_BIDOO);
      	nvgBeginPath(args.vg);
        nvgRoundedRect(args.vg,0,0, box.size.x * std::min((float)module->dataToDecodeRingBuffer.size()/DATASIZE,1.0f),5.f,0.0f);
        nvgRoundedRect(args.vg,0,15.f,box.size.x * std::min((float)module->dataAudioRingBuffer.size()/AUDIOSIZE,1.0f),5.f,0.0f);
      	nvgClosePath(args.vg);
        nvgStroke(args.vg);
      	nvgFill(args.vg);
      	nvgRestore(args.vg);
      }
  	}
  	Widget::drawLayer(args, layer);
  }

};

template <typename BASE>
struct ANTNLight : BASE {
	ANTNLight() {
		this->box.size = mm2px(Vec(6.f, 6.f));
	}
};

struct ANTNWidget : BidooWidget {
	ANTNWidget(ANTN *module) {
		setModule(module);
    prepareThemes(asset::plugin(pluginInstance, "res/ANTN.svg"));

		addChild(createWidget<ScrewSilver>(Vec(15, 0)));
		addChild(createWidget<ScrewSilver>(Vec(box.size.x-30, 0)));
		addChild(createWidget<ScrewSilver>(Vec(15, 365)));
		addChild(createWidget<ScrewSilver>(Vec(box.size.x-30, 365)));

		ANTNBufferDisplay *display = new ANTNBufferDisplay();
		display->module = module;
		display->box.pos = Vec(10.0f, 140.0f);
		display->box.size = Vec(115.0f, 20.0f);
		addChild(display);


    ANTNDisplay* urlDisplay = createWidget<ANTNDisplay>(Vec(5, 25));
		urlDisplay->box.size = Vec(125, 100);
		urlDisplay->setModule(module);
		addChild(urlDisplay);

    addParam(createParam<BidooBlueKnob>(Vec(52.5f, 183), module, ANTN::GAIN_PARAM));

    addParam(createParam<LEDBezel>(Vec(56.5f, 246), module, ANTN::TRIG_PARAM));
		addChild(createLight<ANTNLight<RedGreenBlueLight>>(Vec(58.3f, 247.8f), module, ANTN::ON_LIGHT));

  	static const float portX0[4] = {34, 67, 101};

  	addOutput(createOutput<TinyPJ301MPort>(Vec(portX0[1]-18, 340), module, ANTN::OUTL_OUTPUT));
  	addOutput(createOutput<TinyPJ301MPort>(Vec(portX0[1]+4, 340), module, ANTN::OUTR_OUTPUT));
  }
};

Model *modelANTN = createModel<ANTN, ANTNWidget>("antN");
